src/main/java/org/alfresco/repo/security/authentication/InMemoryTicketComponentImpl.java

```diff
@@ -554,7 +609,26 @@ public boolean equals(Object o)
                 return false;
             }
             Ticket t = (Ticket) o;
-            return (this.expires == t.expires) && this.expiryDate.equals(t.expiryDate) && this.userName.equals(t.userName) && this.ticketId.equals(t.ticketId);
+
+            return this.ticketId.equals(t.ticketId) && this.userName.equals(t.userName) && this.expires.equals(t.expires) && areTheExpiryDatesEqual(t);
+        }
+
+        private boolean areTheExpiryDatesEqual(Ticket t)
+        {
+            if (ExpiryMode.DO_NOT_EXPIRE.equals(this.expires))
+            {
+                // if we have tickets that do not expire, we don't care about their expiration date
+                // this.expiryDate should be null
+                return true;
+            }
+
+            // in this case this.expiryDate should not ever be null; There are checks in the constructors that validate this;
+            // but just in case, somehow it is null: print a warning, then fail
+            if (this.expiryDate == null)
+            {
+                logger.warn("expiryDate should not be null in this case");
+            }
```

Comment:
The method should return here because the following .equals will throw a npe, unless that is expected.

Reply:
I did leave that potential NPE intentionally there (intention stated in the comment above "but just in case, somehow it is null: print a warning, then fail"). 
For the case where you have a ticket with the ExpiryMode set to AFTER_FIXED_TIME (or AFTER_INACTIVITY) but no date to compare to, I couldn't think of a good reason to evaluate the "equals" method and not throw that exception.
In my opinion, we should fail here, as the object that we are testing the equality against is not a valid Ticket object at that stage.
I would have liked to add a junit for this case, but I can't do that without exposing the expiryDate - and I don't want to expose that.

Reply:
Got it.